File: LRUCache.java

import java.util.HashMap;
import java.util.Map;

public class LRUCache {

    Map<Integer, Node<Integer, Integer>> quickAccess_nodeByKey;
    LRUQueue<Integer, Integer> lruQueue;
    int capacity;

    public LRUCache(int capacity) {
        quickAccess_nodeByKey = new HashMap<>();
        lruQueue = new LRUQueue<>();
        this.capacity = capacity;
    }

    public int get(int key) {
        if (quickAccess_nodeByKey.containsKey(key)) {
            lruQueue.removeNode(quickAccess_nodeByKey.get(key));
            lruQueue.addLast(quickAccess_nodeByKey.get(key));
            return quickAccess_nodeByKey.get(key).value;
        }
        return -1;
    }

    public void put(int key, int value) {

        if (quickAccess_nodeByKey.containsKey(key)) {
            lruQueue.removeNode(quickAccess_nodeByKey.get(key));
            lruQueue.addLast(quickAccess_nodeByKey.get(key));
            quickAccess_nodeByKey.get(key).value = value;
            return;
        }

        if (quickAccess_nodeByKey.size() == this.capacity) {
            quickAccess_nodeByKey.remove(lruQueue.head.next.key);
            lruQueue.removeFirst();
        }

        quickAccess_nodeByKey.put(key, new Node(key, value));
        lruQueue.addLast(quickAccess_nodeByKey.get(key));
    }
}

class LRUQueue<K, V> {

    Node<K, V> head;
    Node<K, V> tail;

    public LRUQueue() {
        this.head = new Node<>();
        this.tail = new Node<>();
        this.head.next = this.tail;
        this.tail.previous = this.head;
    }

    public void addLast(Node node) {
        node.next = this.tail;
        node.previous = this.tail.previous;

        this.tail.previous.next = node;
        this.tail.previous = node;
    }

    public void removeFirst() {
        this.head.next = this.head.next.next;
        this.head.next.previous = this.head;
    }

    public void removeNode(Node node) {
        node.next.previous = node.previous;
        node.previous.next = node.next;
    }
}

class Node<K, V> {

    K key;
    V value;
    Node<K, V> next;
    Node<K, V> previous;

    public Node() {
    }

    public Node(K key, V value) {
        this.key = key;
        this.value = value;
    }
}

File: LRUCache.cpp

#include<unordered_map>
using namespace std;

template <typename K, typename V> struct Node {
    K key;
    V value;
    Node<K, V>* next;
    Node<K, V>* previous;

    Node<K, V>() {
        key = -1;
        value = -1;
        next = nullptr;
        previous = nullptr;
    }

    Node(K k, V v) {
        key = k;
        value = v;
        next = nullptr;
        previous = nullptr;
    }
};

template <typename K, typename V> class LRUQueue {
public:
    Node<K, V>* head;
    Node<K, V>* tail;

    LRUQueue() {
        head = new Node<K, V>();
        tail = new Node<K, V>();
        head->next = tail;
        tail->previous = head;
    }

    void addLast(Node<K, V>* node) {
        node->next = tail;
        node->previous = tail->previous;

        tail->previous->next = node;
        tail->previous = node;
    }

    void removeFirst() {
        Node<K, V>* temp = head->next->next;
        delete head->next;
        head->next = temp;
        head->next->previous = head;
        temp = nullptr;
        delete temp;
    }

    void removeNode(Node<K, V>* node) {
        node->next->previous = node->previous;
        node->previous->next = node->next;
        node = nullptr;
        delete node;
    }
};

class LRUCache {
public:
    unordered_map<int, Node<int, int>*> quickAccess_nodeByKey;
    LRUQueue<int, int> lruQueue;
    int max_capacity;

    LRUCache(int capacity) {
        max_capacity = capacity;
    }

    int get(int key) {
        if (quickAccess_nodeByKey.find(key) != quickAccess_nodeByKey.end()) {
            lruQueue.removeNode(quickAccess_nodeByKey[key]);
            lruQueue.addLast(quickAccess_nodeByKey[key]);
            return (quickAccess_nodeByKey[key])->value;
        }
        return -1;
    }

    void put(int key, int value) {
        if (quickAccess_nodeByKey.find(key) != quickAccess_nodeByKey.end()) {
            lruQueue.removeNode(quickAccess_nodeByKey[key]);
            quickAccess_nodeByKey[key]->value = value;
            lruQueue.addLast(quickAccess_nodeByKey[key]);
            return;
        }

        if (quickAccess_nodeByKey.size() == max_capacity) {
            quickAccess_nodeByKey.erase(lruQueue.head->next->key);
            lruQueue.removeFirst();
        }

        quickAccess_nodeByKey[key] = new Node(key, value);
        lruQueue.addLast(quickAccess_nodeByKey[key]);
    }
};
